src/clr-core/internal/decorators/query-slot.ts

```diff
@@ -20,21 +20,42 @@ const standardQuery = (descriptor: PropertyDescriptor, element: any) => ({
   descriptor,
 });
 
+export interface QuerySlotConfig {
+  required?: 'error' | 'warning';
+  requiredMessage?: string;
+}
+
 /**
  * A property decorator that converts a class property into a getter that
  * executes a querySelector on the element's light DOM Slot.
  *
  * @ExportDecoratedItems
  */
-export function querySlot(selector: string) {
+export function querySlot(selector: string, config?: QuerySlotConfig) {
   return (
     protoOrDescriptor: {} | any,
     // tslint:disable-next-line:no-any decorator
     name?: PropertyKey
   ): any => {
     const descriptor = {
       get(this: LitElement) {
-        return this.querySelector(selector);
+        const ref = this.querySelector(selector);
+
+        if (config) {
+          const message = config.requiredMessage
+            ? config.requiredMessage
+            : `The <${selector}> element is required to use <${this.tagName.toLocaleLowerCase()}>`;
+
+          if (!ref && config.required === 'error') {
+            throw new Error(message);
+          }
+
+          if (!ref && config.required === 'warning') {
```

Comment:
so if `config.required` comes in as anything other than `warning` or `error` no message is shown? i know that we are using typescript but i don't think that prevents endusers from making a mistake. could we simplify this a bit and make `warning` the fallback? in the event that a value is passed but isn't expected?

```
if (!ref && config.required) {
  if (config.required === 'error') {
    throw new Error(message);
  } else {
    console.warn(message);
  }
}
```